File: frontend/src/types/stats.ts
export interface TimelinePoint {
  date: string;
  completed: number;
  failed: number;
  overdue: number;
  cancelled: number;
  net_exp: number;
}

export interface OutcomeBreakdown {
  completed: number;
  failed: number;
  overdue: number;
  cancelled: number;
}

export interface PendingSnapshot {
  todo: number;
  in_progress: number;
  on_hold: number;
  rescheduled: number;
}

export interface TimeOfDayDistribution {
  hour: number;
  count: number;
}

export interface HeatmapDay {
  date: string;
  count: number;
  intensity: number;
}

export interface ComparisonMetric {
  current: number;
  previous: number;
  change_percent: number;
  trend: 'up' | 'down' | 'neutral';
}

export interface PeriodComparison {
  period_label: string;
  tasks_completed: ComparisonMetric;
  net_exp: ComparisonMetric;
  completion_rate: ComparisonMetric;
  average_daily_tasks: ComparisonMetric;
}

export interface StatsData {
  account_created_at: string;
  account_age_days: number;
  timeline_granularity: 'day' | 'month';
  timeline: TimelinePoint[];
  outcome_breakdown: OutcomeBreakdown;
  pending_snapshot: PendingSnapshot;
  activity_heatmap: HeatmapDay[];
  time_of_day: TimeOfDayDistribution[];
  total_tasks_completed: number;
  net_exp_change: number;
  completion_rate: number;
  average_tasks_per_day: number;
  most_productive_hour: number | null;
  current_streak: number;
  longest_streak: number;
  monthly_comparison: PeriodComparison | null;
  yearly_comparison: PeriodComparison | null;
}

export type Period = '7d' | '30d' | '90d' | 'all';

File: backend/schemas/stats.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict


class TimelinePoint(BaseModel):
    date: date
    completed: int
    failed: int
    overdue: int
    cancelled: int
    net_exp: int


class OutcomeBreakdown(BaseModel):
    completed: int
    failed: int
    overdue: int
    cancelled: int


class PendingSnapshot(BaseModel):
    todo: int
    in_progress: int
    on_hold: int
    rescheduled: int


class TimeOfDayDistribution(BaseModel):
    hour: int
    count: int


class HeatmapDay(BaseModel):
    date: date
    count: int
    intensity: int


class ComparisonMetric(BaseModel):
    current: float
    previous: float
    change_percent: float
    trend: Literal["up", "down", "neutral"]


class PeriodComparison(BaseModel):
    period_label: str
    tasks_completed: ComparisonMetric
    net_exp: ComparisonMetric
    completion_rate: ComparisonMetric
    average_daily_tasks: ComparisonMetric


class StatsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    account_created_at: date
    account_age_days: int
    timeline_granularity: Literal["day", "month"]
    timeline: list[TimelinePoint]
    outcome_breakdown: OutcomeBreakdown
    pending_snapshot: PendingSnapshot
    activity_heatmap: list[HeatmapDay]
    time_of_day: list[TimeOfDayDistribution]
    total_tasks_completed: int
    net_exp_change: int
    completion_rate: float
    average_tasks_per_day: float
    most_productive_hour: int | None
    current_streak: int
    longest_streak: int
    monthly_comparison: PeriodComparison | None
    yearly_comparison: PeriodComparison | None
